pkg/tfbridge/provider.go

```diff
@@ -408,10 +409,14 @@ func (p *Provider) CheckConfig(ctx context.Context, req *pulumirpc.CheckRequest)
 	// See pulumi/pulumi-terraform-bridge#1087
 	if !news.ContainsUnknowns() {
 		if err := p.preConfigureCallback(ctx, news, config); err != nil {
-			return nil, err
+			return &pulumirpc.CheckResponse{
```

Comment:
This crosses the wires a bit I think.

1. this change removes the ability of the PreConfigureCallback* to return an error; I think this was reasonable to continue being able to return errors, indicating something is going wrong not necessarily user-recoverable or annotated 

2. adaptCheckFailures specifically has heuristics to parse out and failures coming from TF frameworks. It's not obvious why user code from PreConfigureCallback is expected to produce such errors? Does it? Like diagnostic.ValidationError for example?

3. it seems that you would like to add an ability for the PreConfigureCallback to emit CheckFailures. Looks like at some point we moved this callback to execute before CheckConfig instead Configure which makes it a misnomer (I think that hasty move is on me, failed to think this out, but here we go), and currently this signature is this:

```go 
// PreConfigureCallbackWithLogger is a function to invoke prior to calling the T
type PreConfigureCallbackWithLogger func(
	ctx context.Context,
	host *provider.HostClient, vars resource.PropertyMap,
	config shim.ResourceConfig,
) error
```
 
So now that this callback operates on CheckConfig you may be asking for an API to emit CheckFaiures. We could add something opt-in for this. For example we could add:

```
EmitCheckFailure(context.Context, CheckFailure)
``` 

Then PreConfigureCallbackWithLogger that wants to do this specifically would call this API. Without changing any signature here or rerouting errors.

Reply:
I've added an `EmitCheckFailure` function as you suggested - the implementer can call it during PreConfigureCallback and it will append to a list stored in the provider struct.

After the callbacks, we check if it's non-empty and clear it and return the failures.

LMK if that makes sense.